MessageExpression for ValidatingAdmissionPolicy

jiahuif opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
This PR adds support for `messageExpression` to ValidatingAdmissionPolicy, to set a custom message when the policy rejects a request.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
*API REVIEW* Please take a look at the first commit labeled `[API REVIEW]`
Please note that the API is mirrored from https://github.com/kubernetes/kubernetes/pull/115969, which already got API approved.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added `messageExpression` to `ValidatingAdmissionPolicy`, to set custom failure message via CEL expression.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jiahuif committed:
[API REVIEW] Validation.MessageExpression

jiahuif committed:
implement message expression.

cici37 commented:
/triage accepted
/assign

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -119,24 +122,24 @@ func objectToResolveVal(r runtime.Object) (interface{}, error) {
 	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
```
maybe return -1 for cost when err occur?

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -160,6 +163,9 @@ func (f *filter) ForInput(ctx context.Context, versionedAttr *admission.Versione
 
 	remainingBudget := runtimeCELCostBudget
 	for i, compilationResult := range f.compilationResults {
+		if compilationResult.Program == nil {
```
Program is nil could due to some err in compilationResult.Error...

jiahuif replied:
This case was checked before the code reaches here. nil Problem can only mean a nil input in the same index.

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -119,24 +122,24 @@ func objectToResolveVal(r runtime.Object) (interface{}, error) {
 	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
```
Same for all following(-1)?

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -119,24 +122,24 @@ func objectToResolveVal(r runtime.Object) (interface{}, error) {
 // errors per evaluation are returned on the Evaluation object
 // runtimeCELCostBudget was added for testing purpose only. Callers should always use const RuntimeCELCostBudget from k8s.io/apiserver/pkg/apis/cel/config.go as input.
-func (f *filter) ForInput(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *admissionv1.AdmissionRequest, inputs OptionalVariableBindings, runtimeCELCostBudget int64) ([]EvaluationResult, error) {
+func (f *filter) ForInput(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *admissionv1.AdmissionRequest, inputs OptionalVariableBindings, runtimeCELCostBudget int64) ([]EvaluationResult, int64, error) {
```
It might be introduced by my earlier pr but: Should we check `runtimeCELCostBudget` early and if it is <0 stop immediately?

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go:
```diff
@@ -101,30 +105,45 @@ func (v *validator) Validate(ctx context.Context, versionedAttr *admission.Versi
 			decision.Evaluation = EvalError
 			decision.Message = evalResult.Error.Error()
+		} else if messageResult != nil {
+			err := messageResult.Error
```
If internal err happens to messageExpr, we will return as well

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go:
```diff
@@ -101,30 +105,45 @@ func (v *validator) Validate(ctx context.Context, versionedAttr *admission.Versi
-				decision.Message = strings.TrimSpace(validation.Message)
+			message := ""
+			if messageResult != nil && messageResult.Error == nil && messageResult.EvalResult != nil {
+				message = messageResult.EvalResult.Value().(string)
```
check if messageExpression evaluated to "" or nil? (in this case should fall back to use `message` as well)

cici37 replied:
Maybe check if it resolved to `string` type first?

jiahuif replied:
update the code to reflect the new behaviour.

cici37 reviewed (commented):
Thank you for raising this PR! I know you might have started to work on it already but just a reminder of the work/tests:
1. API validation on `messageExpression` and it's tests.(We do compilation on API validation level hence will need support for new added `messageExpression` field as well)
2. Tests on combination of all possible exists status for `messageExpression`, `message`, `expression`; 
3. `messageExpression` fail to compile
4. `messageExpression` fail to eval to string
5.  err handling behavior.
6. runtime cost exceed for `messageExpression`
7. overall budget exceed for `messageExpression`
8. If the cost calculated successfully for `messageExpression`

Thank you so much for the work! :)

cici37 commented:
/milestone v1.27

jiahuif commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -256,6 +256,16 @@ type Validation struct {
+	// If the evaluation fails for whatever reason, Message will be displayed instead.
+	// MessageExpression has access to the following variables:
+	// object, oldObject, params, and request
+	// +optional
```
Noted. Will sync with said PR.

cici37 replied:
Sorry forgot to mention that we'd better use the serialized field names in godoc which is `messageExpression` and also we will check if line breaks in messageExpression. Here is the updated doc:
```
// messageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails.
	// Since messageExpression is used as a failure message, it must evaluate to a string.
	// If both message and messageExpression are present on a validation, then messageExpression will be used if validation fails. 
	// If messageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
	// as if the messageExpression field were unset. If messageExpression evaluates to an empty string, a string with only spaces, or a string
	// that contains line breaks, then the validation failure message will also be produced as if the messageExpression field were unset, and
	// the fact that messageExpression produced an empty string/string with only spaces/string with line breaks will be logged.
	// messageExpression has access to all the same variables as the `expression` except for 'authorizer' and 'authorizer.requestResource'.
	// Example:
	// "object.x must be less than max ("+string(self.max)+")"
	// +optional
```
So we have to check the linebreaks in expressionExpr as what we did here:
https://github.com/kubernetes/kubernetes/blob/c80a93d1a9fb62f62acd923be63124116bde3f58/pkg/apis/admissionregistration/validation/validation.go#L808-L809
Inside the same func, check the messageExpression successfully compiled as well

jiahuif replied:
Since, #115969 is merged, I am resolving this thread after last sync.

rjsadow commented:
/release-note-edit

```release-note Added `MessageExpression` to `ValidatingAdmissionPolicy`, to set custom failure message via CEL expression.```

cici37 commented:
> /release-note-edit
> 
> `` release-note Added `MessageExpression` to `ValidatingAdmissionPolicy`, to set custom failure message via CEL expression. ``

Thanks @rjsadow for the update! Lets change `MessageExpression` to `messageExpression` to match the actual field name. @jiahuif

jiahuif commented:
- [ ] API validation on messageExpression and it's tests.(We do compilation on API validation level hence will need support for new added messageExpression field as well)
- [ ] Tests on combination of all possible exists status for messageExpression, message, expression;
- [ ] messageExpression fail to compile
- [ ] messageExpression fail to eval to string
- [ ] err handling behavior.
- [ ] runtime cost exceed for messageExpression
- [ ] overall budget exceed for messageExpression
- [ ] If the cost calculated successfully for messageExpression

jpbetz commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -256,6 +256,18 @@ type Validation struct {
+	// the fact that messageExpression produced an empty string/string with only spaces/string with line breaks will be logged.
+	// messageExpression has access to all the same variables as the `expression` except for 'authorizer' and 'authorizer.requestResource'.
+	// Example:
+	// "object.x must be less than max ("+string(self.max)+")"
```
s/self/object to match the variables available in this API

cici37 replied:
Let's switch `self` to `params` here. Sorry my earlier comment provided for doc forgot to update this

jiahuif replied:
Updated.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go:
```diff
@@ -101,30 +106,57 @@ func (v *validator) Validate(ctx context.Context, versionedAttr *admission.Versi
 
 	options := cel.OptionalVariableBindings{VersionedParams: versionedParams}
-	auditAnnotationEvalResults, err := v.auditAnnotationFilter.ForInput(ctx, versionedAttr, cel.CreateAdmissionRequest(versionedAttr.Attributes), options, runtimeCELCostBudget)
+	auditAnnotationEvalResults, _, err := v.auditAnnotationFilter.ForInput(ctx, versionedAttr, cel.CreateAdmissionRequest(versionedAttr.Attributes), options, runtimeCELCostBudget)
```
I realized when reading this that I didn't share the budget between `validations` and `auditAnnoations`.  BTH, I didn't think it through carefully.  Any thoughts on if they should be kept separate or combined?  I'm currently in favor of keeping them separate (the way the code exists now) since `auditAnnotations` are logically their own capability separate from validation, but I'm curious what others think.

jiahuif replied:
@cici37 We had a discussion on this topic. Since it was a bit rushed we did not have a good idea about sharing cost vs dedicated cost. Could you help me answer this concern? Thank you.

cici37 replied:
After offline discussion with @jpbetz , let's have the runtimeCELCostBudget only for validations which is shared by expression and messageExpression. auditAnnotations and possible matchConditions will have separate budgets.

jiahuif replied:
This behavior is consistent with current implementation.

jpbetz commented:
Looking good! Once https://github.com/kubernetes/kubernetes/pull/115969 merged and all API impacting changes are sync'd to this, LGTM.

valaparthvi commented:
Hi @jiahuif, I am from the Bug Triage Team for K8s release v1.27. I see this PR was opened yday and is in actively worked up, so I assume it is on track for this release.

However, please note that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

wenjiaswe commented:
> Hi Jiahui Feng, I am from the Bug Triage Team for K8s release v1.27. I see this PR was opened yday and is in actively worked up, so I assume it is on track for this release.
> 
> However, please note that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

@valaparthvi Thank you very much for your reminding. Sure, everything is ready, this will be in hopefully tomorrow.

jiahuif committed:
generated: ./hack/update-codegen.sh && ./hack/update-openapi-spec.sh

jiahuif commented:
/retest

jiahuif commented:
/label api-review
Please note that the API is mirrored from https://github.com/kubernetes/kubernetes/pull/115969, which already got API approved. I hope this note expedites the API review.

jiahuif commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -209,6 +209,16 @@ type Validation struct {
+	// If the evaluation fails for whatever reason, Message will be displayed instead.
+	// MessageExpression has access to the following variables:
+	// object, oldObject, params, and request
+	// +optional
```
Since #115969 is merged, I am resolving this thread after last sync.

cici37 commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -785,23 +785,16 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
+	if len(v.MessageExpression) != 0 {
+		allErrors = append(allErrors, validateCELExpression(v.MessageExpression, plugincel.OptionalVariableDeclarations{
+			HasParams:     paramKind != nil,
+			HasAuthorizer: true,
```
This should be false.

cici37 commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -785,23 +785,16 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
+			HasAuthorizer: true,
+		}, fldPath.Child("expression"))...)
+	}
+	if len(v.MessageExpression) != 0 {
```
Do we wanna add a check of `if len(v. MessageExpression) > 0 && len(trimmedMessageExpression) == 0` here?

jiahuif replied:
If it contains only whitespaces, we should treat as an empty expression, and go ahead to expression validation. Is thi correct/

jpbetz replied:
Consistency with the above handling of `Expression` is probably good here.

cici37 replied:
I was thinking to handle it same as for v.Message:
```
if len(v.Message) > 0 && len(trimmedMsg) == 0 {
		allErrors = append(allErrors, field.Invalid(fldPath.Child("message"), v.Message, "message must be non-empty if specified"))
```

jiahuif replied:
That's fixed. However we need to leave out the field name in the message, as it automatically filled.

cici37 commented on pkg/apis/admissionregistration/validation/validation_test.go:
```diff
@@ -2564,7 +2564,38 @@ func TestValidateValidatingAdmissionPolicy(t *testing.T) {
 				},
 			},
-			expectedError: `spec.validations[0].expression: Invalid value: "object.x in [1, 2, ": compilation failed: ERROR: <input>:1:19: Syntax error: missing ']' at '<EOF>`,
+			expectedError: `spec.validations[0].expression: Invalid value: "object.x in [1, 2, ": compilation failed: ERROR: <input>:1:20: Syntax error: missing ']' at '<EOF>`,
```
It makes sense. Thank you!

jiahuif committed:
generated: UPDATE_COMPATIBILITY_FIXTURE_DATA

(cd staging/src/k8s.io/api/ && env UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test)

jiahuif commented:
/retest

liggitt commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -782,26 +782,26 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
+		}, fldPath.Child("expression"))...)
+	}
+	if len(v.MessageExpression) > 0 && len(trimmedMessageExpression) == 0 {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("message"), v.Message, "must be non-empty if specified"))
```
```suggestion
		allErrors = append(allErrors, field.Invalid(fldPath.Child("messageExpression"), v.MessageExpression, "must be non-empty if specified"))
```

liggitt commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -782,26 +782,26 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
+	}
+	if len(v.MessageExpression) > 0 && len(trimmedMessageExpression) == 0 {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("message"), v.Message, "must be non-empty if specified"))
+	} else if hasNewlines(trimmedMessageExpression) {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("message"), v.Message, "must not contain line breaks"))
```
drop this, we don't care if the messageExpression CEL has linebreaks, we care about the string it produces not having linebreaks

jiahuif replied:
fixed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -78,6 +78,9 @@ func (a *evaluationActivation) Parent() interpreter.Activation {
 func (c *filterCompiler) Compile(expressionAccessors []ExpressionAccessor, options OptionalVariableDeclarations, perCallLimit uint64) Filter {
 	compilationResults := make([]CompilationResult, len(expressionAccessors))
 	for i, expressionAccessor := range expressionAccessors {
+		if expressionAccessor == nil {
```
this is surprising... is leaving zero-value entries in compilationResults going to break expectations anywhere we send this list?

cici37 replied:
More info: This case is made special for messageExpression only since it could be unset and we wanna it to be a 1:1 mapping to validations[i].expression. For the rest of cases, the cel expression is expected to be always non-nil(It is guaranteed from api level and code impl).

jiahuif replied:
Thank you @cici37 for explanation. Here this special handling is an unfortunate compromise of the interface changes made in https://github.com/kubernetes/kubernetes/pull/115816 . Allowing nil as placeholders is least impactful on the ongoing development like https://github.com/kubernetes/kubernetes/pull/116261 that rely on the interfaces. I will send another refactor once all affecting PRs are merged, after this release.

liggitt replied:
ok, as long as you've swept call sites / implementations to make sure they're all expecting the nil items, a refactor follow-up can be ok

jiahuif replied:
Yes these are all manually examined and fixed.

cici37 replied:
Could we add test case for this?

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller_reconcile.go:
```diff
@@ -459,9 +459,11 @@ func (c *policyController) latestPolicyData() []policyData {
 					hasParam = true
 				}
 				optionalVars := cel.OptionalVariableDeclarations{HasParams: hasParam, HasAuthorizer: true}
+				expressionOptionalVars := cel.OptionalVariableDeclarations{HasParams: true, HasAuthorizer: false}
```
```suggestion
				expressionOptionalVars := cel.OptionalVariableDeclarations{HasParams: hasParam, HasAuthorizer: false}
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller_reconcile.go:
```diff
@@ -514,6 +516,19 @@ func convertv1alpha1Validations(inputValidations []v1alpha1.Validation) []cel.Ex
+			validation := MessageExpressionCondition{
+				MessageExpression: validation.MessageExpression,
+			}
+			celExpressionAccessor[i] = &validation
```
nit: taking the address of the `validation` loop var in a way that escapes the function reads like a bug until you realize `validation` is shadowed... pick a different name for the var to avoid that confusion?

jiahuif replied:
Fixed. A previous reviewer suggested to make this function consistent with the other conversion functions below, which all have this problem.

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller_reconcile.go:
```diff
@@ -514,6 +516,19 @@ func convertv1alpha1Validations(inputValidations []v1alpha1.Validation) []cel.Ex
+func convertV1Alpha1MessageExpressions(inputValidations []v1alpha1.Validation) []cel.ExpressionAccessor {
+	celExpressionAccessor := make([]cel.ExpressionAccessor, len(inputValidations))
+	for i, validation := range inputValidations {
+		if validation.MessageExpression != "" {
```
anywhere this is "" we're leaving nil entries in the returned celExpressionAccessor list... that's really subtle and could easily lead to NPE bugs

jiahuif replied:
This is a compromise of a refactor that other ongoing development relies on (see reply above). This is the least intrusive way to do placeholders. I will send another refactor once affected developments finish.

liggitt commented:
just a couple tweaks in the API validation, then lgtm. I noticed a few things in the impl but didn't review that closely, leaning on @cici37 and @jpbetz for that review

liggitt commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -782,26 +782,24 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
+		}, fldPath.Child("expression"))...)
+	}
+	if len(v.MessageExpression) > 0 && len(trimmedMessageExpression) == 0 {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("messageExpression"), v.Message, "must be non-empty if specified"))
```
```suggestion
		allErrors = append(allErrors, field.Invalid(fldPath.Child("messageExpression"), v.MessageExpression, "must be non-empty if specified"))
```

jiahuif replied:
oops. mispushed on this part. fixed.

liggitt commented:
/approve
for API changes
@cici37 or @jpbetz have lgtm for message impl

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go:
```diff
@@ -92,9 +92,10 @@ type OptionalVariableBindings struct {
 	// runtimeCELCostBudget was added for testing purpose only. Callers should always use const RuntimeCELCostBudget from k8s.io/apiserver/pkg/apis/cel/config.go as input.
-	ForInput(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *v1.AdmissionRequest, optionalVars OptionalVariableBindings, runtimeCELCostBudget int64) ([]EvaluationResult, error)
+	// If cost budget is calculated, the filter should return the remaining budget.
+	ForInput(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *v1.AdmissionRequest, optionalVars OptionalVariableBindings, runtimeCELCostBudget int64) ([]EvaluationResult, int64, error)
```
Add test case for returned remaining budget?

jiahuif replied:
added.

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator_test.go:
```diff
@@ -43,11 +44,17 @@ type fakeCelFilter struct {
 
-func (f *fakeCelFilter) ForInput(context.Context, *admission.VersionedAttributes, *admissionv1.AdmissionRequest, cel.OptionalVariableBindings, int64) ([]cel.EvaluationResult, error) {
+func (f *fakeCelFilter) ForInput(_ context.Context, _ *admission.VersionedAttributes, _ *admissionv1.AdmissionRequest, _ cel.OptionalVariableBindings, costBudget int64) ([]cel.EvaluationResult, int64, error) {
+	if costBudget <= 0 {
```
<0?

jiahuif replied:
Not really. the expression itself costs one. Needs at least 1 to cover it.

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator_test.go:
```diff
@@ -43,11 +44,17 @@ type fakeCelFilter struct {
+	}
 	if f.throwError {
-		return nil, errors.New("test error")
+		return nil, 0, errors.New("test error")
```
-1?

jiahuif replied:
fixed. Thanks.

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator_test.go:
```diff
@@ -572,6 +581,244 @@ func TestValidate(t *testing.T) {
+			},
+		},
+		{
+			name: "messageExpression out of budget after successful eval of expression",
```
Great to see the test for budget shared between expression and messageExpression

jiahuif replied:
:)

cici37 commented:
/lgtm
/approve

Thank you!

jiahuif commented:
Thank you all for reviewing.
